### ISSUE TITLE
[MPS] Multilayer LSTM backward ignores dropout, silently corrupting training gradients

### ISSUE THREAD

**Issue by bogatyy** [pytorch/pytorch#190056]
### 🐛 Describe the bug

`nn.LSTM` with `num_layers >= 2` and `dropout > 0` computes wrong gradients on
MPS. The forward saves each layer's output *before* applying inter-layer
dropout, and the backward consumes those unmasked tensors and never reapplies
the mask — so both the inter-layer gradient chain and the upper layers' weight
gradients are computed as if no dropout had happened.

Deterministic check via fixed-RNG central differences (resetting the device
RNG before every forward makes the dropout mask identical across evaluations):

```python
import torch

torch.manual_seed(0)
lstm = torch.nn.LSTM(4, 4, num_layers=2, dropout=0.5).to("mps")
lstm.train()

def loss(x):
    torch.mps.manual_seed(42)   # identical dropout mask on every call
    out, _ = lstm(x)
    return out.sum()

x = torch.randn(3, 2, 4, device="mps", requires_grad=True)
loss(x).backward()
analytic = x.grad.flatten()[0].item()
eps = 1e-2
with torch.no_grad():
    xp = x.detach().clone(); xp.flatten()[0] += eps
    xm = x.detach().clone(); xm.flatten()[0] -= eps
    numeric = ((loss(xp) - loss(xm)) / (2 * eps)).item()
print(f"analytic={analytic:.6f} numeric={numeric:.6f}")
```

Output on a source build of current main (2.14.0a0+git2662808); also
reproduces on 2.13.0 stable and nightly 2.14.0.dev20260710:

```text
analytic=0.000132 numeric=0.006306
```

That is a 98% relative error. Running the same script with `dropout=0.0` as a
control gives `analytic=-0.007924 numeric=-0.007927` (0.04% error), so the
finite-difference setup itself is sound. Weight gradients are similarly wrong
(a two-layer `dropout=0.5` LSTM had layer-0 weight gradient analytic -9.69e-4
vs numeric +1.97e-4).

The forward stores pre-dropout layer outputs in `layersOutputs`
(https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/mps/operations/RnnOps.mm#L219-L256),
and the backward slices them as the lower layers' inputs and calls
`LSTMGradientsWithSourceTensor` with `mask:nil`, with no mask multiply in the
inter-layer gradient chain
(https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/mps/operations/RnnOps.mm#L580-L615).

### Versions

PyTorch 2.14.0a0+git2662808 (source build of main), 2.13.0, and nightly 2.14.0.dev20260710; macOS 26.5.2 (arm64), Apple M5 Max, Python 3.12.13.
